plugins/cache/bap_cache.ml

```diff
@@ -0,0 +1,178 @@
+open Core_kernel
+open Regular.Std
+open Bap.Std
+open Bap_cache_types
+
+include Self ()
+
+module Filename = Caml.Filename
+module Utils = Bap_cache_utils
+
+let (//) = Filename.concat
+
+module Cfg = struct
+
+  module T = struct
+    type t = config [@@deriving bin_io, compare, sexp]
+  end
+
+  let default = {
+    capacity = 4 * 1024; (* 4 Gb  *)
+    overhead = 0.25;
+    gc_enabled = true;
+  }
+
+  let version = 3
+  let config_file = sprintf "config.%d" version
+  let cache_data  = "data"
+
+  let config_file path = path // config_file
+  let data  path = path // cache_data
+
+  let gc_threshold c =
+    c.capacity + Int.of_float (float c.capacity *. c.overhead)
+
+  let default_root = ref None
+
+  let set_root dir = default_root := Some dir
+
+  let write path cfg =
+    try Utils.binable_to_file (module T) path cfg
+    with e ->
+      warning "storing config: %s" (Exn.to_string e)
+
+  let read path =
+    try Utils.binable_from_file (module T) path
+    with e ->
+      warning "read config: %s" (Exn.to_string e);
+      default
+
+end
+
+include Cfg
+
+let getenv opt = try Some (Sys.getenv opt) with Caml.Not_found -> None
+
+let root () =
+  let root = match !default_root with
+    | Some dir -> dir // ".cache" // "bap"
+    | None -> match getenv "XDG_CACHE_HOME" with
+      | Some cache -> cache
+      | None -> match getenv "HOME" with
+        | None -> Filename.get_temp_dir_name () // "bap" // "cache"
+        | Some home -> home // ".cache" // "bap" in
+  root
+
+let ensure_dir_exists path =
+  try
+    Unix.mkdir path 0o700
+  with
+  | Unix.(Unix_error (EEXIST,_,_)) -> ()
+  | exn -> raise exn
+
+let rec mkdir path =
+  let par = Filename.dirname path in
+  if not (Sys.file_exists par) then mkdir par;
+  if not (Sys.file_exists path) then
+    ensure_dir_exists path
+
+let dir_exists dir = Sys.file_exists dir && Sys.is_directory dir
+
+let with_temp_dir path ~f =
+  let tmp = Filename.temp_file "tmp" "" in
```

Comment:
This is very fragile, `temp_file @@ mkdir` is not an atomic operation, not even close, as there is no even a check that the folder/file exists. Therefore, it will be quite often when two or more different processes will try to create a folder with the same name. To make it more robust you shall do the following:
```
1. choose a new random name
2. create a folder with that name, if not successful goto 1.
```
The operating system will guarantee that the created folder, if successful was created by the current process.